react/features/base/redux/StateListenerRegistry.js

```diff
@@ -0,0 +1,136 @@
+     *
+     * @param {Object} context - The redux store invoking the listener and the
+     * private state of this {@code StateListenerRegistry} associated with the
+     * redux store.
+     * @returns {void}
+     */
+    _listener({ prevSelections, store }) {
+        for (const selectorListener of this._selectorListeners) {
+            const prevSelection = prevSelections[selectorListener];
+
+            try {
+                const selection
+                    = selectorListener.selector(
+                        store.getState(),
+                        prevSelection);
+
+                if (prevSelection !== selection) {
+                    prevSelections[selectorListener] = selection;
+                    selectorListener.listener(selection, store, prevSelection);
+                }
+            } catch (e) {
+                // Don't let one faulty listener prevent other listeners from
+                // being notified about their associated changes.
+                logger.error(e);
+            }
+        }
+    }
+
+    /**
+     * Registers a specific listener to be notified when the value derived by a
+     * specific {@code selector} from a redux store/state changes.
+     *
+     * @param {Function} selector - The pure {@code Function} of the redux
+     * store/state (and the previous selection of made by {@code selector})
+     * which selects the value listened to by the specified {@code listener}.
+     * @param {Function} listener - The listener to register with this
+     * {@code StateListenerRegistry} so that it gets invoked when the value
+     * returned by the specified {@code selector} changes.
+     * @returns {void}
+     */
+    register(selector: Selector, listener: Listener) {
+        this._selectorListeners.add({
+            listener,
+            selector
+        });
+    }
+
+    /**
+     * Subscribes to a specific redux store (so that this instance gets notified
+     * any time an action is dispatched, and some part of the state (tree) of
+     * the specified redux store may potentially have changed).
+     *
+     * @param {Store} store - The redux store to which this
+     * {@code StateListenerRegistry} is to {@code subscribe}.
+     * @returns {void}
+     */
+    subscribe(store: Store) {
+        // XXX If StateListenerRegistry is not utilized by the app to listen to
+        // state changes, do not bother subscribing to the store at all.
+        if (this._selectorListeners.size) {
+            store.subscribe(
+                this._listener.bind(
+                    this,
+                    {
+                        /**
+                         * The previous selections of the {@code Selector}s
+                         * registered with this {@code StateListenerRegistry}.
+                         *
+                         * @type Map<any>
+                         */
+                        prevSelections: new Map(),
```

Comment:
Obviously, I didn't change the uses of `prevSelections` to use a `Map` and they rather add properties to it. But that's immaterial to the abstraction right now and I'll fix it later on should be decide to go with it.